### PR TITLE
SITES-784: Better error handling. Skip stanford:* type workgroups.

### PR DESCRIPTION
**Pull request by sherakama**
# READY FOR REVIEW

# Summary
- Better error handling for bad rules
- Better logging and handling for bad API requests
- Handling for `stanford:*` type workgroup maps. These don't work in the API.
- Force SAML attribute affiliation mapping to handle the `stanford:*` mappings instead. 
- SAML attribute affiliation mapping on user register.

# Needed By (Date)
- For Greg

# Urgency
- Ask Greg

# Steps to Test

1. Copy dennylab to a test environment
2. Log in with sso/login and view logs
3. See PHP errors and warnings 
4. Check out this branch
5. Log out and log back in
6. See no more PHP errors.
7. Delete user account
8. Log out
9. Log back in
10. Validate Affiliation roles have been assigned

# Affected Projects or Products
- All of them
- SOE Migration

# Associated Issues and/or People
- @boznik 
- @jbickar 
- SITES-784

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

**Comment by jbickar** [comment]
Good catch. The following are not workgroups, but rather "privgroups":
* `stanford:staff`
* `stanford:faculty`
* `stanford:student`
* `stanford:all`

It is possible that there may actually be _workgroups_ under the `stanford:` stem. I am not sure. We should find out.

**Comment by sherakama** [comment]
We had discussions on this with the MAIS team and a little light bulb is shining where I remember they mentioned that they were special. The workgroup API returns a 404 when trying to use them as a role mapping.

**Comment by jbickar** [comment]
https://uit.stanford.edu/service/saml/arp

**Comment by jbickar** [comment]
I cannot find anything in Workgroup Manager when searching on the `stanford:` stem. That does not mean that there aren't any, but it's a good indication.

**Comment by jbickar** [comment]
You couldn't convince @boznik to go with SAML `eduPersonEntitlement` workgroup mapping, huh? I tried my damnedest.

**Comment by jbickar** [comment]
TL;DR we need to ask the MAIS team.

**Comment by sherakama** [comment]
Yeah, we have the eduPersonAffiliation that we can rely on for this kind of mapping. I am fine with not mapping `stanford:*` but the module will have to account for those rules that were converted from Webauth.

**Comment by jbickar** [comment]
Yeah, it's going to have to convert those to `eduPersonAffiliation`. Ugh.

**Comment by sherakama** [comment]
I'm fine with keeping the scope of this module to the 'known' affiliate attributes and keeping private workgroups private. I think that makes sense in the scope of this. We should provide a path for exposing the private groups at special requests but I am leaning towards leaving that kind of work for D8.

**Comment by jbickar** [comment]
The private workgroups issue is outside of the scope of this PR. We can burn that bridge when we get there.

**Comment by boznik** [comment]
thx for going the extra mile on this...it's all working now.

**Comment by jbickar** [comment]
Those who are tardy do not get fruit cup. Nor do they get to review PRs, apparently.
